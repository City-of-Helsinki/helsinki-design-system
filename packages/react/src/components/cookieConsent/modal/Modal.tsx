import React, { useContext, useEffect, useState } from 'react';
import { VisuallyHidden } from '@react-aria/visually-hidden';

import classNames from '../../../utils/classNames';
import styles from '../CookieConsent.module.scss';
import { CookieConsentContext, useCookieConsentUiTexts } from '../CookieConsentContext';
import Content from '../content/Content';

export function Modal(): React.ReactElement | null {
  const cookieConsentContext = useContext(CookieConsentContext);
  const shouldShowModal = !cookieConsentContext.hasUserHandledAllConsents();
  const [isModalInitiallyShown] = useState<boolean>(shouldShowModal);
  const [popupTimerComplete, setPopupTimerComplete] = useState<boolean>(false);
  const popupDelayInMs = 500;
  // if hasUserHandledAllConsents() was false at first and then later true, user must have saved consents.
  const showScreenReaderSaveNotification = isModalInitiallyShown && !shouldShowModal;
  const { settingsSaved } = useCookieConsentUiTexts();
  useEffect(() => {
    setTimeout(() => setPopupTimerComplete(true), popupDelayInMs);
  }, []);

  if (showScreenReaderSaveNotification) {
    return (
      <VisuallyHidden>
        <div role="alert" data-testid="cookie-consent-screen-reader-notification" tabIndex={-1}>
          {settingsSaved}
        </div>
      </VisuallyHidden>
    );
  }

  if (!shouldShowModal) {
    return null;
  }

  return (
    <div
      className={classNames(styles.container, popupTimerComplete && styles['animate-in'])}
      data-testid="cookie-consent"
    >
      <div className={styles.aligner}>{popupTimerComplete && <Content />}</div>
    </div>
  );
}
