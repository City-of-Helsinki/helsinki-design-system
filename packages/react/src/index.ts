export * from './icons';
export * from './components';
export * from './ssr';
