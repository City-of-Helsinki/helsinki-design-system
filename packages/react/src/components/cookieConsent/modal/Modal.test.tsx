/* eslint-disable jest/expect-expect */
/* eslint-disable jest/no-mocks-import */
import React from 'react';
import { render, RenderResult, waitFor } from '@testing-library/react';
import { axe } from 'jest-axe';
import { act } from 'react-dom/test-utils';

import { Modal } from './Modal';
import { COOKIE_NAME } from '../cookieConsentController';
import { Content, Provider as CookieContextProvider } from '../CookieConsentContext';
import mockDocumentCookie from '../__mocks__/mockDocumentCookie';
import {
  clickElement,
  extractSetCookieArguments,
  getContent,
  isAccordionOpen,
  openAccordion,
  verifyElementDoesNotExistsByTestId,
  verifyElementExistsByTestId,
  commonTestProps,
  TestConsentData,
  TestGroupParent,
  createCookieDataWithSelectedRejections,
} from '../test.util';

const { requiredGroupParent, optionalGroupParent, defaultConsentData, unknownConsents, dataTestIds } = commonTestProps;

const mockedCookieControls = mockDocumentCookie();

let content: Content;

const renderCookieConsent = (
  { requiredConsents = [], optionalConsents = [], cookie = {}, contentModifier }: TestConsentData,
  withRealTimers = false,
): RenderResult => {
  // inject unknown consents to verify those are
  // stored and handled, but not required or optional
  const cookieWithInjectedUnknowns = {
    ...cookie,
    ...unknownConsents,
  };
  content = getContent(requiredConsents, optionalConsents, contentModifier);
  jest.useFakeTimers();
  mockedCookieControls.init({ [COOKIE_NAME]: JSON.stringify(cookieWithInjectedUnknowns) });
  const result = render(
    <CookieContextProvider content={content}>
      <Modal />
    </CookieContextProvider>,
  );
  act(() => {
    jest.runAllTimers();
  });

  // For example, axe uses timers so sometimes the test must use real ones
  if (withRealTimers) {
    jest.useRealTimers();
  }

  return result;
};

describe('<ModalContent /> spec', () => {
  afterEach(() => {
    mockedCookieControls.clear();
  });

  afterAll(() => {
    mockedCookieControls.restore();
  });

  it('renders the component', () => {
    const { asFragment } = renderCookieConsent(defaultConsentData);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', async () => {
    const { container } = renderCookieConsent(defaultConsentData, true);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe('<CookieConsent /> ', () => {
  afterEach(() => {
    mockedCookieControls.clear();
  });

  afterAll(() => {
    mockedCookieControls.restore();
  });

  const getSetCookieArguments = (index = -1) => extractSetCookieArguments(mockedCookieControls, index);

  const initDetailsView = async (data: TestConsentData): Promise<RenderResult> => {
    const result = renderCookieConsent(data);
    await openAccordion(result, dataTestIds.settingsToggler);
    return result;
  };

  describe('Cookie consent ', () => {
    it('and child components are rendered when consents have not been handled', () => {
      const result = renderCookieConsent(defaultConsentData);
      verifyElementExistsByTestId(result, dataTestIds.container);
      verifyElementExistsByTestId(result, dataTestIds.languageSwitcher);
      verifyElementExistsByTestId(result, dataTestIds.approveButton);
      verifyElementExistsByTestId(result, dataTestIds.approveRequiredButton);
    });

    it('is rendered if a required consent has not been approved. It could have been optional before', () => {
      const result = renderCookieConsent({
        ...defaultConsentData,
        cookie: createCookieDataWithSelectedRejections(['requiredConsent2']),
      });
      verifyElementExistsByTestId(result, dataTestIds.container);
    });

    it('is not shown when all consents have been handled and are true/false', () => {
      const result = renderCookieConsent({
        ...defaultConsentData,
        cookie: createCookieDataWithSelectedRejections(['optionalConsent1', 'optionalConsent3']),
      });
      verifyElementDoesNotExistsByTestId(result, dataTestIds.container);
      verifyElementDoesNotExistsByTestId(result, dataTestIds.screenReaderNotification);
    });
    it('changing language calls content.onLanguageChange', () => {
      const onLanguageChange = jest.fn();
      const result = renderCookieConsent({
        ...defaultConsentData,
        contentModifier: (currentContent) => {
          // eslint-disable-next-line no-param-reassign
          currentContent.language.onLanguageChange = onLanguageChange;
          return currentContent;
        },
      });
      result.container.querySelector('#cookie-consent-language-selector-button').click();
      result.container.querySelector('a[lang="sv"]').click();
      expect(onLanguageChange).toHaveBeenLastCalledWith('sv');
    });
  });

  describe(`Approve button will 
            - hide the cookie consent
            - show a prompt for screen readers
            - save cookie`, () => {
    const checkCookiesAreSetAndConsentModalHidden = (result: RenderResult, assumedConsents: unknown) => {
      expect(JSON.parse(getSetCookieArguments().data)).toEqual(assumedConsents);
      verifyElementDoesNotExistsByTestId(result, dataTestIds.container);
      verifyElementExistsByTestId(result, dataTestIds.screenReaderNotification);
      expect(mockedCookieControls.mockSet).toHaveBeenCalledTimes(1);
    };

    it('Approve -button approves all consents when details are not shown', () => {
      const result = renderCookieConsent(defaultConsentData);
      const consentResult = {
        ...createCookieDataWithSelectedRejections([]),
        ...unknownConsents,
      };
      clickElement(result, dataTestIds.approveButton);
      checkCookiesAreSetAndConsentModalHidden(result, consentResult);
    });

    it('Approve required -button approves only required consents and clears selected consents', async () => {
      const result = renderCookieConsent(defaultConsentData);
      const consentResult = {
        ...createCookieDataWithSelectedRejections(['optionalConsent1', 'optionalConsent2', 'optionalConsent3']),
        ...unknownConsents,
      };
      await openAccordion(result, dataTestIds.settingsToggler);
      clickElement(result, dataTestIds.getConsentGroupCheckboxId(optionalGroupParent, 0));
      clickElement(result, dataTestIds.getConsentGroupCheckboxId(optionalGroupParent, 1));
      clickElement(result, dataTestIds.approveRequiredButton);
      checkCookiesAreSetAndConsentModalHidden(result, consentResult);
    });

    it('Approve -button will approve required and selected consents when details are shown', async () => {
      const result = renderCookieConsent(defaultConsentData);
      const consentResult = {
        ...createCookieDataWithSelectedRejections(['optionalConsent2', 'optionalConsent3']),
        ...unknownConsents,
      };
      await openAccordion(result, dataTestIds.settingsToggler);
      clickElement(result, dataTestIds.getConsentGroupCheckboxId(optionalGroupParent, 0));
      clickElement(result, dataTestIds.approveButton);
      checkCookiesAreSetAndConsentModalHidden(result, consentResult);
    });
  });

  describe('Settings accordion can be opened and in details view ', () => {
    it('required and optional consent groups are rendered', async () => {
      const result = await initDetailsView(defaultConsentData);
      verifyElementExistsByTestId(result, dataTestIds.getConsentsCheckboxId(requiredGroupParent));
      verifyElementExistsByTestId(result, dataTestIds.getConsentsCheckboxId(optionalGroupParent));
      defaultConsentData.requiredConsents.forEach((consent, index) => {
        verifyElementExistsByTestId(result, dataTestIds.getConsentGroupCheckboxId(requiredGroupParent, index));
        verifyElementExistsByTestId(result, dataTestIds.getConsentGroupDetailsTogglerId(requiredGroupParent, index));
      });
      defaultConsentData.optionalConsents.forEach((consent, index) => {
        verifyElementExistsByTestId(result, dataTestIds.getConsentGroupCheckboxId(optionalGroupParent, index));
        verifyElementExistsByTestId(result, dataTestIds.getConsentGroupDetailsTogglerId(optionalGroupParent, index));
      });
    });

    it('Approve button text changes when accordion is open vs closed', async () => {
      const result = await initDetailsView(defaultConsentData);
      const approveButtonTextWhileOpen = (result.getByTestId(dataTestIds.approveButton) as HTMLElement).innerHTML;

      clickElement(result, dataTestIds.settingsToggler);
      await waitFor(() => {
        expect(isAccordionOpen(result, dataTestIds.settingsToggler)).toBeFalsy();
      });
      const approveButtonTextWhileClosed = (result.getByTestId(dataTestIds.approveButton) as HTMLElement).innerHTML;
      expect(approveButtonTextWhileOpen).not.toBe(approveButtonTextWhileClosed);
    });

    it(`clicking an optional consent group sets the all consents in that group true/false. 
        Cookie consent is not hidden until an approve -button is clicked`, async () => {
      const result = await initDetailsView(defaultConsentData);
      defaultConsentData.optionalConsents.forEach((consent, index) => {
        clickElement(result, dataTestIds.getConsentGroupCheckboxId(optionalGroupParent, index));
      });
      clickElement(result, dataTestIds.getConsentGroupCheckboxId(optionalGroupParent, 0));
      clickElement(result, dataTestIds.approveButton);
      expect(JSON.parse(getSetCookieArguments().data)).toEqual({
        ...createCookieDataWithSelectedRejections(['optionalConsent1']),
        ...unknownConsents,
      });
      verifyElementDoesNotExistsByTestId(result, dataTestIds.container);
      verifyElementExistsByTestId(result, dataTestIds.screenReaderNotification);
      expect(mockedCookieControls.mockSet).toHaveBeenCalledTimes(1);
    });
  });
  describe('Accordions of each consent group can be opened and ', () => {
    it('all consents in the group are rendered', async () => {
      const result = await initDetailsView(defaultConsentData);
      const checkConsentsExist = async (groupParent: TestGroupParent) => {
        const list =
          groupParent === 'required' ? content.requiredConsents.groupList : content.optionalConsents.groupList;
        let index = 0;
        // cannot use async/await with array.forEach
        // eslint-disable-next-line no-restricted-syntax
        for (const groups of list) {
          expect(result.getByTestId(dataTestIds.getConsentGroupTableId(groupParent, index))).not.toBeVisible();
          // eslint-disable-next-line no-await-in-loop
          await openAccordion(result, dataTestIds.getConsentGroupDetailsTogglerId(groupParent, index));
          expect(result.getByTestId(dataTestIds.getConsentGroupTableId(groupParent, index))).toBeVisible();
          index += 1;
          groups.consents.forEach((consent) => {
            expect(result.getAllByText(consent.name)).toHaveLength(1);
          });
        }
      };
      await checkConsentsExist('required');
      await checkConsentsExist('optional');
    });
  });
});
