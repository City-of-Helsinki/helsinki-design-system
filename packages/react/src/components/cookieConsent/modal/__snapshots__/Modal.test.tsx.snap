// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalContent /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="container modal animate-in"
    data-testid="cookie-consent"
  >
    <div
      class="aligner"
    >
      <div
        class="content shrink-on-blur"
        id="cookie-consent-content"
        tabindex="-1"
      >
        <div
          class="main-content"
          data-testid="cookie-consent-information"
        >
          <span
            aria-level="1"
            class="emulated-h1 visually-hidden-without-focus"
            role="heading"
            tabindex="-1"
          >
            Test site eväste­suostumukset
          </span>
          <div
            class="language-switcher"
            data-testid="cookie-consent-language-switcher"
          >
            <div
              class="menuButton language-selector-override"
            >
              <button
                aria-controls="cookie-consent-language-selector-menu"
                aria-expanded="false"
                aria-haspopup="true"
                aria-label="Kieli: Suomi. Vaihda kieli. Change language. Ändra språk."
                class="toggleButton"
                id="cookie-consent-language-selector-button"
                type="button"
              >
                <span
                  class="toggleButtonLabel"
                >
                  Suomeksi (FI)
                </span>
                <svg
                  aria-hidden="true"
                  class="icon s"
                  role="img"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    fill="none"
                    fill-rule="evenodd"
                  >
                    <path
                      d="M0 0h24v24H0z"
                    />
                    <path
                      d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                      fill="currentColor"
                    />
                  </g>
                </svg>
              </button>
              <div
                aria-labelledby="cookie-consent-language-selector-button"
                class="menu"
                id="cookie-consent-language-selector-menu"
                role="region"
                style="top: 10px; min-width: 190px;"
              >
                <a
                  aria-current="page"
                  class="item "
                  href="#"
                  lang="fi"
                >
                  <span
                    class="label"
                  >
                    Suomeksi (FI)
                  </span>
                </a>
                <a
                  class="item "
                  href="#"
                  lang="sv"
                >
                  <span
                    class="label"
                  >
                    Svenska (SV)
                  </span>
                </a>
                <a
                  class="item "
                  href="#"
                  lang="en"
                >
                  <span
                    class="label"
                  >
                    English (EN)
                  </span>
                </a>
              </div>
            </div>
          </div>
          <p
            class="cut-text-without-focus"
          >
            Tämä sivusto käyttää välttämättömiä evästeitä suorituskyvyn varmistamiseksi sekä yleisen käytön seurantaan. Lisäksi käytämme kohdennusevästeitä käyttäjäkokemuksen parantamiseksi, analytiikkaan ja kohdistetun sisällön näyttämiseen. Jatkamalla sivuston käyttöä ilman asetusten muuttamista hyväksyt välttämättömien evästeiden käytön.
          </p>
          <button
            class="accordion-button shown-without-focus"
            type="button"
          >
            Lue lisää
          </button>
        </div>
        <button
          aria-expanded="false"
          class="accordion-button accordion-button-settings-closed hidden-without-focus"
          data-testid="cookie-consent-settings-toggler"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="icon s"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M0 0h24v24H0z"
              />
              <path
                d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                fill="currentColor"
              />
            </g>
          </svg>
          <span>
            Näytä asetukset
          </span>
        </button>
        <div
          class="card custom-theme-1"
          role="region"
          style="display: none;"
        >
          <div
            class="text-content"
            data-testid="cookie-consent-details"
          >
            <span
              aria-level="2"
              class="emulated-h2"
              role="heading"
            >
              Tietoa sivustolla käytetyistä evästeistä
            </span>
            <p>
              Sivustolla käytetyt evästeet on luokiteltu käyttötarkoituksen mukaan. Alla voit lukea tietoa jokaisesta kategoriasta ja sallia tai kieltää evästeiden käytön.
            </p>
            <div
              class="consent-group-parent"
            >
              <div
                class="title-with-checkbox"
              >
                <div
                  class="checkbox"
                  style="--label-font-size: var(--fontsize-heading-m);"
                >
                  <input
                    aria-describedby="required-consents-description"
                    checked=""
                    class="input"
                    data-testid="required-consents-checkbox"
                    disabled=""
                    id="required-consents-checkbox"
                    name="required-consents-checkbox"
                    type="checkbox"
                    value=""
                  />
                  <label
                    class="label"
                    for="required-consents-checkbox"
                  >
                    Title for required cookies
                  </label>
                </div>
              </div>
              <p
                aria-hidden="true"
              >
                Text for required cookies
              </p>
              <div
                aria-hidden="true"
                id="required-consents-description"
                style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
              >
                checkboxAriaLabel
              </div>
              <ul
                class="list"
              >
                <li>
                  <div
                    class="consent-group"
                  >
                    <div
                      class="title-with-checkbox"
                    >
                      <div
                        class="checkbox"
                        style="--label-font-size: var(--fontsize-heading-s);"
                      >
                        <input
                          aria-describedby="required-consents-group-0-description"
                          checked=""
                          class="input"
                          data-testid="required-consents-group-0-checkbox"
                          disabled=""
                          id="required-consents-group-0-checkbox"
                          name="required-consents-group-0-checkbox"
                          type="checkbox"
                          value=""
                        />
                        <label
                          class="label"
                          for="required-consents-group-0-checkbox"
                        >
                          Cookie group title for requiredCookieGroup0
                        </label>
                      </div>
                    </div>
                    <div
                      class="consent-group-content"
                    >
                      <p
                        aria-hidden="true"
                      >
                        Cookie group description for requiredCookieGroup0
                      </p>
                      <div
                        aria-hidden="true"
                        id="required-consents-group-0-description"
                        style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
                      >
                        checkboxAriaLabel for requiredCookieGroup0
                      </div>
                      <button
                        aria-expanded="false"
                        aria-label="expandAriaLabel for requiredCookieGroup0"
                        class="accordion-button"
                        data-testid="required-consents-group-0-details-toggler"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="icon s"
                          role="img"
                          viewBox="0 0 24 24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <g
                            fill="none"
                            fill-rule="evenodd"
                          >
                            <path
                              d="M0 0h24v24H0z"
                            />
                            <path
                              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                              fill="currentColor"
                            />
                          </g>
                        </svg>
                      </button>
                      <div
                        aria-label="Cookie group title for requiredCookieGroup0"
                        class="card custom-theme-2"
                        role="region"
                        style="display: none;"
                      >
                        <div
                          class="data-table-container"
                        >
                          <div
                            class="container"
                            tabindex="0"
                          >
                            <table
                              class="table dark dense custom-theme-3"
                              data-testid="required-consents-group-0-table"
                              id="required-consents-group-0-table"
                            >
                              <thead>
                                <tr
                                  class="headerRow"
                                >
                                  <th
                                    scope="col"
                                  >
                                    Nimi
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Osoite
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Kuvaus
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Voimassaoloaika
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class="content"
                              >
                                <tr>
                                  <td
                                    data-testid="name-0"
                                  >
                                    Name of requiredConsent1
                                  </td>
                                  <td
                                    data-testid="hostName-0"
                                  >
                                    HostName of requiredConsent1
                                  </td>
                                  <td
                                    data-testid="description-0"
                                  >
                                    Description of requiredConsent1
                                  </td>
                                  <td
                                    data-testid="expiration-0"
                                  >
                                    Expiration of requiredConsent1
                                  </td>
                                </tr>
                                <tr>
                                  <td
                                    data-testid="name-1"
                                  >
                                    Name of requiredConsent2
                                  </td>
                                  <td
                                    data-testid="hostName-1"
                                  >
                                    HostName of requiredConsent2
                                  </td>
                                  <td
                                    data-testid="description-1"
                                  >
                                    Description of requiredConsent2
                                  </td>
                                  <td
                                    data-testid="expiration-1"
                                  >
                                    Expiration of requiredConsent2
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
                <li>
                  <div
                    class="consent-group"
                  >
                    <div
                      class="title-with-checkbox"
                    >
                      <div
                        class="checkbox"
                        style="--label-font-size: var(--fontsize-heading-s);"
                      >
                        <input
                          aria-describedby="required-consents-group-1-description"
                          checked=""
                          class="input"
                          data-testid="required-consents-group-1-checkbox"
                          disabled=""
                          id="required-consents-group-1-checkbox"
                          name="required-consents-group-1-checkbox"
                          type="checkbox"
                          value=""
                        />
                        <label
                          class="label"
                          for="required-consents-group-1-checkbox"
                        >
                          Cookie group title for requiredCookieGroup1
                        </label>
                      </div>
                    </div>
                    <div
                      class="consent-group-content"
                    >
                      <p
                        aria-hidden="true"
                      >
                        Cookie group description for requiredCookieGroup1
                      </p>
                      <div
                        aria-hidden="true"
                        id="required-consents-group-1-description"
                        style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
                      >
                        checkboxAriaLabel for requiredCookieGroup1
                      </div>
                      <button
                        aria-expanded="false"
                        aria-label="expandAriaLabel for requiredCookieGroup1"
                        class="accordion-button"
                        data-testid="required-consents-group-1-details-toggler"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="icon s"
                          role="img"
                          viewBox="0 0 24 24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <g
                            fill="none"
                            fill-rule="evenodd"
                          >
                            <path
                              d="M0 0h24v24H0z"
                            />
                            <path
                              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                              fill="currentColor"
                            />
                          </g>
                        </svg>
                      </button>
                      <div
                        aria-label="Cookie group title for requiredCookieGroup1"
                        class="card custom-theme-4"
                        role="region"
                        style="display: none;"
                      >
                        <div
                          class="data-table-container"
                        >
                          <div
                            class="container"
                            tabindex="0"
                          >
                            <table
                              class="table dark dense custom-theme-5"
                              data-testid="required-consents-group-1-table"
                              id="required-consents-group-1-table"
                            >
                              <thead>
                                <tr
                                  class="headerRow"
                                >
                                  <th
                                    scope="col"
                                  >
                                    Nimi
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Osoite
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Kuvaus
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Voimassaoloaika
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class="content"
                              >
                                <tr>
                                  <td
                                    data-testid="name-0"
                                  >
                                    Name of requiredConsent3
                                  </td>
                                  <td
                                    data-testid="hostName-0"
                                  >
                                    HostName of requiredConsent3
                                  </td>
                                  <td
                                    data-testid="description-0"
                                  >
                                    Description of requiredConsent3
                                  </td>
                                  <td
                                    data-testid="expiration-0"
                                  >
                                    Expiration of requiredConsent3
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
            <div
              class="consent-group-parent"
            >
              <div
                class="title-with-checkbox"
              >
                <div
                  class="checkbox"
                  style="--label-font-size: var(--fontsize-heading-m);"
                >
                  <input
                    aria-describedby="optional-consents-description"
                    class="input"
                    data-testid="optional-consents-checkbox"
                    id="optional-consents-checkbox"
                    name="optional-consents-checkbox"
                    type="checkbox"
                    value=""
                  />
                  <label
                    class="label"
                    for="optional-consents-checkbox"
                  >
                    Title for optional cookies
                  </label>
                </div>
              </div>
              <p
                aria-hidden="true"
              >
                Text for optional cookies
              </p>
              <div
                aria-hidden="true"
                id="optional-consents-description"
                style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
              >
                checkboxAriaLabel
              </div>
              <ul
                class="list"
              >
                <li>
                  <div
                    class="consent-group"
                  >
                    <div
                      class="title-with-checkbox"
                    >
                      <div
                        class="checkbox"
                        style="--label-font-size: var(--fontsize-heading-s);"
                      >
                        <input
                          aria-describedby="optional-consents-group-0-description"
                          class="input"
                          data-testid="optional-consents-group-0-checkbox"
                          id="optional-consents-group-0-checkbox"
                          name="optional-consents-group-0-checkbox"
                          type="checkbox"
                          value=""
                        />
                        <label
                          class="label"
                          for="optional-consents-group-0-checkbox"
                        >
                          Cookie group title for optionalCookieGroups0
                        </label>
                      </div>
                    </div>
                    <div
                      class="consent-group-content"
                    >
                      <p
                        aria-hidden="true"
                      >
                        Cookie group description for optionalCookieGroups0
                      </p>
                      <div
                        aria-hidden="true"
                        id="optional-consents-group-0-description"
                        style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
                      >
                        checkboxAriaLabel for optionalCookieGroups0
                      </div>
                      <button
                        aria-expanded="false"
                        aria-label="expandAriaLabel for optionalCookieGroups0"
                        class="accordion-button"
                        data-testid="optional-consents-group-0-details-toggler"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="icon s"
                          role="img"
                          viewBox="0 0 24 24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <g
                            fill="none"
                            fill-rule="evenodd"
                          >
                            <path
                              d="M0 0h24v24H0z"
                            />
                            <path
                              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                              fill="currentColor"
                            />
                          </g>
                        </svg>
                      </button>
                      <div
                        aria-label="Cookie group title for optionalCookieGroups0"
                        class="card custom-theme-6"
                        role="region"
                        style="display: none;"
                      >
                        <div
                          class="data-table-container"
                        >
                          <div
                            class="container"
                            tabindex="0"
                          >
                            <table
                              class="table dark dense custom-theme-7"
                              data-testid="optional-consents-group-0-table"
                              id="optional-consents-group-0-table"
                            >
                              <thead>
                                <tr
                                  class="headerRow"
                                >
                                  <th
                                    scope="col"
                                  >
                                    Nimi
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Osoite
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Kuvaus
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Voimassaoloaika
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class="content"
                              >
                                <tr>
                                  <td
                                    data-testid="name-0"
                                  >
                                    Name of optionalConsent1
                                  </td>
                                  <td
                                    data-testid="hostName-0"
                                  >
                                    HostName of optionalConsent1
                                  </td>
                                  <td
                                    data-testid="description-0"
                                  >
                                    Description of optionalConsent1
                                  </td>
                                  <td
                                    data-testid="expiration-0"
                                  >
                                    Expiration of optionalConsent1
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
                <li>
                  <div
                    class="consent-group"
                  >
                    <div
                      class="title-with-checkbox"
                    >
                      <div
                        class="checkbox"
                        style="--label-font-size: var(--fontsize-heading-s);"
                      >
                        <input
                          aria-describedby="optional-consents-group-1-description"
                          class="input"
                          data-testid="optional-consents-group-1-checkbox"
                          id="optional-consents-group-1-checkbox"
                          name="optional-consents-group-1-checkbox"
                          type="checkbox"
                          value=""
                        />
                        <label
                          class="label"
                          for="optional-consents-group-1-checkbox"
                        >
                          Cookie group title for optionalCookieGroups1
                        </label>
                      </div>
                    </div>
                    <div
                      class="consent-group-content"
                    >
                      <p
                        aria-hidden="true"
                      >
                        Cookie group description for optionalCookieGroups1
                      </p>
                      <div
                        aria-hidden="true"
                        id="optional-consents-group-1-description"
                        style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
                      >
                        checkboxAriaLabel for optionalCookieGroups1
                      </div>
                      <button
                        aria-expanded="false"
                        aria-label="expandAriaLabel for optionalCookieGroups1"
                        class="accordion-button"
                        data-testid="optional-consents-group-1-details-toggler"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="icon s"
                          role="img"
                          viewBox="0 0 24 24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <g
                            fill="none"
                            fill-rule="evenodd"
                          >
                            <path
                              d="M0 0h24v24H0z"
                            />
                            <path
                              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                              fill="currentColor"
                            />
                          </g>
                        </svg>
                      </button>
                      <div
                        aria-label="Cookie group title for optionalCookieGroups1"
                        class="card custom-theme-8"
                        role="region"
                        style="display: none;"
                      >
                        <div
                          class="data-table-container"
                        >
                          <div
                            class="container"
                            tabindex="0"
                          >
                            <table
                              class="table dark dense custom-theme-9"
                              data-testid="optional-consents-group-1-table"
                              id="optional-consents-group-1-table"
                            >
                              <thead>
                                <tr
                                  class="headerRow"
                                >
                                  <th
                                    scope="col"
                                  >
                                    Nimi
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Osoite
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Kuvaus
                                  </th>
                                  <th
                                    scope="col"
                                  >
                                    Voimassaoloaika
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class="content"
                              >
                                <tr>
                                  <td
                                    data-testid="name-0"
                                  >
                                    Name of optionalConsent2
                                  </td>
                                  <td
                                    data-testid="hostName-0"
                                  >
                                    HostName of optionalConsent2
                                  </td>
                                  <td
                                    data-testid="description-0"
                                  >
                                    Description of optionalConsent2
                                  </td>
                                  <td
                                    data-testid="expiration-0"
                                  >
                                    Expiration of optionalConsent2
                                  </td>
                                </tr>
                                <tr>
                                  <td
                                    data-testid="name-1"
                                  >
                                    Name of optionalConsent3
                                  </td>
                                  <td
                                    data-testid="hostName-1"
                                  >
                                    HostName of optionalConsent3
                                  </td>
                                  <td
                                    data-testid="description-1"
                                  >
                                    Description of optionalConsent3
                                  </td>
                                  <td
                                    data-testid="expiration-1"
                                  >
                                    Expiration of optionalConsent3
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="buttons"
        >
          <button
            class="button primary theme-default size-default"
            data-testid="cookie-consent-approve-button"
            type="button"
          >
            <span
              class="label"
            >
              Hyväksy kaikki evästeet
            </span>
          </button>
          <button
            class="button secondary theme-default size-default"
            data-testid="cookie-consent-approve-required-button"
            type="button"
          >
            <span
              class="label"
            >
              Hyväksy vain pakolliset evästeet
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
